apps/wallet-mobile/src/features/Swap/common/SwapProvider.tsx

```diff
@@ -0,0 +1,506 @@
+        break
+      case SwapAction.EstimateResponse:
+        draft.lastInputTouched = state.lastInputTouched
+        draft.reqres = 'response'
+        draft.estimate = action.value
+        draft.tokenOutInput.error = null
+        draft.canSwap = true
+        draft.wantedPrice = state.wantedPrice === '' ? String(action.value.netPrice) : state.wantedPrice
+        if (state.lastInputTouched === 'in') {
+          draft.tokenOutInput.value = String(action.value.totalOutputWithoutSlippage ?? 0)
+        } else {
+          draft.tokenInInput.value = String(action.value.totalInput ?? 0)
+        }
+
+        break
+      case SwapAction.EstimateError:
+        draft.estimate = undefined
+        draft.tokenOutInput.error = action.value.message
+        draft.canSwap = false
+
+        break
+      case SwapAction.CreateResponse:
+        draft.createTx = action.value
+
+        break
+      case SwapAction.CreateError:
+        draft.createTx = undefined
+        draft.tokenOutInput.error = action.value.message
+
+        break
+      default:
+        throw new Error(`swapReducer invalid action`)
+    }
+  })
+}
+
+export const SwapAction = {
+  ChangeOrderType: 'ChangeOrderType',
+  TokenInInputTouched: 'TokenInInputTouched',
+  TokenOutInputTouched: 'TokenOutInputTouched',
+  TokenInIdChanged: 'TokenInIdChanged',
+  TokenOutIdChanged: 'TokenOutIdChanged',
+  TokenInAmountChanged: 'TokenInAmountChanged',
+  TokenOutAmountChanged: 'TokenOutAmountChanged',
+  TokenInErrorChanged: 'TokenInErrorChanged',
+  TokenOutErrorChanged: 'TokenOutErrorChanged',
+  WantedPriceInputChanged: 'WantedPriceInputChanged',
+  SlippageInputChanged: 'SlippageInputChanged',
+  SwitchTouched: 'SwitchTouched',
+  DexSelectorTouched: 'DexSelectorTouched',
+  Refresh: 'Refresh',
+  ResetAmounts: 'ResetAmounts',
+  ResetForm: 'ResetForm',
+  EstimateResponse: 'EstimateResponse',
+  EstimateError: 'EstimateError',
+  CreateResponse: 'CreateResponse',
+  CreateError: 'CreateError',
+} as const
+
+type SwapActionValueMap = {
+  ChangeOrderType: 'limit' | 'market'
+  TokenInInputTouched: undefined
+  TokenOutInputTouched: undefined
+  TokenInIdChanged: Portfolio.Token.Id
+  TokenOutIdChanged: Portfolio.Token.Id
+  TokenInAmountChanged: string
+  TokenOutAmountChanged: string
+  TokenInErrorChanged: string | null
+  TokenOutErrorChanged: string | null
+  WantedPriceInputChanged: string
+  SlippageInputChanged: number
+  SwitchTouched: undefined
+  DexSelectorTouched: Swap.Provider
+  Refresh: undefined
+  ResetAmounts: undefined
+  ResetForm: undefined
+  EstimateResponse: Swap.EstimateResponse
+  EstimateError: Api.ResponseError
+  CreateResponse: Swap.CreateResponse
+  CreateError: Api.ResponseError
+}
+
+export type SwapAction = {
+  [K in keyof SwapActionValueMap]: SwapActionValueMap[K] extends undefined
+    ? {type: K}
+    : {type: K; value: SwapActionValueMap[K]}
+}[keyof SwapActionValueMap]
+
+const defaultState: SwapState = Object.freeze({
+  reqres: 'response',
```

Comment:
what is it role `reqres` ?

Reply:
Need some way to differentiate between updates to the state that trigger a new estimate request, or updates to the state that don't trigger an estimate: for example, the estimate response itself, or errors, or the create response, etc.
Could be named better.